[JSC] Add Fuzzilli support to jsc

ddkilzer opened the pull request:
#### 59ee7065d4c9870d8ad953a0b2d34a68f2c25e4e
<pre>
[JSC] Add Fuzzilli support to jsc
<a href="https://bugs.webkit.org/show_bug.cgi?id=264983">https://bugs.webkit.org/show_bug.cgi?id=264983</a>
&lt;<a href="https://rdar.apple.com/118531102">rdar://118531102</a>&gt;

Reviewed by Yusuke Suzuki.

Based on a patch by Samuel Groß and Carl Smith:
&lt;<a href="https://github.com/googleprojectzero/fuzzilli/blob/main/Targets/JavaScriptCore/Patches/webkit.patch">https://github.com/googleprojectzero/fuzzilli/blob/main/Targets/JavaScriptCore/Patches/webkit.patch</a>&gt;

* Configurations/CommonBase.xcconfig:
- Move support for adding -DENABLE_LIBFUZZER=1 to OTHER_CFLAGS from here
  to Sanitizers.xcconfig since LIBFUZZER support is there now.
* Configurations/Sanitizers.xcconfig:
- Add -DENABLE_LIBFUZZER=1 to OTHER_CFLAGS when ENABLE_LIBFUZZER=YES.
- Add -fsanitize-coverage=trace-pc-guard -DENABLE_FUZZILLI=1 to
  OTHER_CFLAGS when ENABLE_FUZZILLI=YES.
- Add -fsanitize-coverage=trace-pc-guard to OTHER_LDFLAGS when when
  ENABLE_FUZZILLI=YES.

* Source/JavaScriptCore/CMakeLists.txt:
- Add fuzzilli to the header search path for CMake builds.
* Source/JavaScriptCore/Configurations/JSC.xcconfig:
(JSC_USE_STATIC_LINKING):
- Set to YES only when ENABLE_FUZZILLI=YES as this is the only
  configuration where it is currently required.
* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig:
- Disable static initializer check at link time when
  ENABLE_FUZZILLI=YES or when WK_ANY_SANITIZER_ENABLED=YES.
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
- Add fuzzilli/Fuzzilli.{cpp,h} to the jsc target in the Xcode project.

* Source/JavaScriptCore/jsc.cpp:
(GlobalObject::promiseRejectionTracker):
(JSC_DEFINE_HOST_FUNCTION):
(runWithOptions):
(printUsageStatement):
(CommandLine::parseArguments):
(runJSC):
- Make changes to support Fuzzilli.  CommandLine::m_repl is always
  defined, but only set to true when Fuzzilli support is enabled.
  This simplifies some logic that would otherwise need to use the
  ENABLE(FUZZILLI) macro.

* Source/JavaScriptCore/fuzzilli/Fuzzilli.cpp: Add.
(Fuzzilli::resetCoverageEdges):
(Fuzzilli::logFile):
(Fuzzilli::waitForCommand):
(Fuzzilli::initializeCoverage):
(Fuzzilli::readInput):
(Fuzzilli::flushReprl):
(Fuzzilli::initializeReprl):
(__sanitizer_cov_trace_pc_guard_init):
(__sanitizer_cov_trace_pc_guard):
* Source/JavaScriptCore/fuzzilli/Fuzzilli.h: Add.
- Move core Fuzzilli code to its own header and source files.

* Tools/Scripts/set-webkit-configuration:
(printCurrentSettings):
- Add support for --[no-]fuzzilli command-line switches.
- Enabling Fuzzilli support also disables libFuzzer support.
* Tools/Scripts/webkitdirs.pm:
(determineFuzzilliIsEnabled):
(fuzzilliIsEnabled):
(XcodeOptions):
- Add support for the --[no-]fuzzilli command-line switches in
  set-webkit-configuration.

Canonical link: <a href="https://commits.webkit.org/270923@main">https://commits.webkit.org/270923@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ee510a545d9c948de2a29e3ec85c2aedcf4210df

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26821 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5437 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28057 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29032 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24514 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27267 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7273 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2828 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24404 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27083 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4249 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23020 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3729 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3775 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24017 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29512 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/23370 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24459 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24419 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30019 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/26008 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3808 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2003 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27927 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5262 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/33461 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4275 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7234 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3476 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4171 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/fuzzilli/Fuzzilli.h:
```diff
@@ -0,0 +1,63 @@
+
+void resetCoverageEdges();
+
+FILE* logFile();
```
Can we use FilePrintStream instead? This can make WebKit classes used without various hacks.

Constellation commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -3189,6 +3226,57 @@ JSC_DEFINE_HOST_FUNCTION(functionDropAllLocks, (JSGlobalObject* globalObject, Ca
+        RETURN_IF_EXCEPTION(scope, { });
+
+        auto cString = toCString(globalObject, scope, string);
+        RETURN_IF_EXCEPTION(scope, { });
```
This is not necessary if we are using FilePrintStream.
Then,
```
Fuzzilli::logFile().println(string);
Fuzzilli::logFile().flush();
```

Constellation commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -3189,6 +3226,57 @@ JSC_DEFINE_HOST_FUNCTION(functionDropAllLocks, (JSGlobalObject* globalObject, Ca
+        // We directly require a string as argument for simplicity
+        return JSValue::encode(jsUndefined());
+    }
+    auto operation = callFrame->argument(0).toString(globalObject)->value(globalObject);
```
Use `toWTFString(globalObject)` instead of `toString(...)->value(...)`

Constellation commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -3189,6 +3226,57 @@ JSC_DEFINE_HOST_FUNCTION(functionDropAllLocks, (JSGlobalObject* globalObject, Ca
+        return JSValue::encode(jsUndefined());
+    }
+    auto operation = callFrame->argument(0).toString(globalObject)->value(globalObject);
+    RETURN_IF_EXCEPTION(scope, encodedJSValue());
```
Use `{ }` instead of `encodedJSValue()` in new code.

Constellation commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -3189,6 +3226,57 @@ JSC_DEFINE_HOST_FUNCTION(functionDropAllLocks, (JSGlobalObject* globalObject, Ca
+
+    if (operation == "FUZZILLI_CRASH"_s) {
+        int32_t command = callFrame->argument(1).toInt32(globalObject);
+        RETURN_IF_EXCEPTION(scope, encodedJSValue());
```
Ditto.

Constellation commented on Source/JavaScriptCore/fuzzilli/Fuzzilli.h:
```diff
@@ -0,0 +1,63 @@
+
+#pragma once
+
+#include <fcntl.h>
```
We can include `<wtf/FilePrintStream.h>` instead if we use it.

Constellation commented on Source/JavaScriptCore/fuzzilli/Fuzzilli.cpp:
```diff
@@ -0,0 +1,171 @@
+            if (!result) {
+                result = stdout;
+
+                fprintf(stderr, "Fuzzer output channel not available, printing to stdout instead\n");
```
Let's use `dataLogLn("Fuzzer output channel not available, printing to stdout instead");`

Constellation commented on Source/JavaScriptCore/fuzzilli/Fuzzilli.cpp:
```diff
@@ -0,0 +1,171 @@
+        sharedData = static_cast<SharedData*>(mmap(0, SHM_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0));
+        RELEASE_ASSERT_WITH_MESSAGE(sharedData != MAP_FAILED, "Failed to mmap shared memory region");
+
+        fprintf(stderr, "[COV] edge counters initialized. Shared memory: %s with %zu edges\n", shmKey, edgesStop - edgesStart);
```
Let's use `dataLogLn("[COV] edge counters initialized. Shared memory: ", shmKey, " with ", edgesStop - edgesStart, " edges");`

Constellation reviewed (approved):
r=me

junghoon4a committed:
[JSC] Add Fuzzilli support to jsc
https://bugs.webkit.org/show_bug.cgi?id=264983
<rdar://118531102>

Reviewed by Yusuke Suzuki.

Based on a patch by Samuel Groß and Carl Smith:
<https://github.com/googleprojectzero/fuzzilli/blob/main/Targets/JavaScriptCore/Patches/webkit.patch>

* Configurations/CommonBase.xcconfig:
- Move support for adding -DENABLE_LIBFUZZER=1 to OTHER_CFLAGS from here
  to Sanitizers.xcconfig since LIBFUZZER support is there now.
* Configurations/Sanitizers.xcconfig:
- Add -DENABLE_LIBFUZZER=1 to OTHER_CFLAGS when ENABLE_LIBFUZZER=YES.
- Add -fsanitize-coverage=trace-pc-guard -DENABLE_FUZZILLI=1 to
  OTHER_CFLAGS when ENABLE_FUZZILLI=YES.
- Add -fsanitize-coverage=trace-pc-guard to OTHER_LDFLAGS when when
  ENABLE_FUZZILLI=YES.

* Source/JavaScriptCore/CMakeLists.txt:
- Add fuzzilli to the header search path for CMake builds.
* Source/JavaScriptCore/Configurations/JSC.xcconfig:
(JSC_USE_STATIC_LINKING):
- Set to YES only when ENABLE_FUZZILLI=YES as this is the only
  configuration where it is currently required.
* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig:
- Disable static initializer check at link time when
  ENABLE_FUZZILLI=YES or when WK_ANY_SANITIZER_ENABLED=YES.
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
- Add fuzzilli/Fuzzilli.{cpp,h} to the jsc target in the Xcode project.

* Source/JavaScriptCore/jsc.cpp:
(GlobalObject::promiseRejectionTracker):
(JSC_DEFINE_HOST_FUNCTION):
(runWithOptions):
(printUsageStatement):
(CommandLine::parseArguments):
(runJSC):
- Make changes to support Fuzzilli.  CommandLine::m_repl is always
  defined, but only set to true when Fuzzilli support is enabled.
  This simplifies some logic that would otherwise need to use the
  ENABLE(FUZZILLI) macro.

* Source/JavaScriptCore/fuzzilli/Fuzzilli.cpp: Add.
(Fuzzilli::resetCoverageEdges):
(Fuzzilli::logFile):
(Fuzzilli::waitForCommand):
(Fuzzilli::initializeCoverage):
(Fuzzilli::readInput):
(Fuzzilli::flushReprl):
(Fuzzilli::initializeReprl):
(__sanitizer_cov_trace_pc_guard_init):
(__sanitizer_cov_trace_pc_guard):
* Source/JavaScriptCore/fuzzilli/Fuzzilli.h: Add.
- Move core Fuzzilli code to its own header and source files.

* Tools/Scripts/set-webkit-configuration:
(printCurrentSettings):
- Add support for --[no-]fuzzilli command-line switches.
- Enabling Fuzzilli support also disables libFuzzer support.
* Tools/Scripts/webkitdirs.pm:
(determineFuzzilliIsEnabled):
(fuzzilliIsEnabled):
(XcodeOptions):
- Add support for the --[no-]fuzzilli command-line switches in
  set-webkit-configuration.

Canonical link: https://commits.webkit.org/270923@main
